dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java

```diff
@@ -0,0 +1,215 @@
+      Session session = sessionPromise.get(5, TimeUnit.SECONDS);
+
+      // Prepare the request
+      HttpFields requestFields = new HttpFields();
+      if (credentials != null) {
+        credentials.getRequestMetadata();
+        requestFields.add(HttpHeader.AUTHORIZATION,
+            "Bearer " + credentials.getAccessToken().getTokenValue());
+      }
+      requestFields.add(HttpHeader.CONTENT_TYPE,
+          "application/dicom");
+      MetaData.Request request = new MetaData.Request("POST", uri, HttpVersion.HTTP_2,
+          requestFields);
+      HeadersFrame headersFrame = new HeadersFrame(request, null, false);
+
+      // Prepare the listener to receive the HTTP response frames.
+      final StringBuilder resultBuilder = new StringBuilder();
+      final CompletableFuture<Integer> responseCodeFuture = new CompletableFuture<>();
+      final CompletableFuture<Boolean> doneFuture = new CompletableFuture<>();
+      Stream.Listener responseListener = new Stream.Listener.Adapter() {
+        @Override
+        public void onReset(Stream stream, ResetFrame frame) {
+          doneFuture.complete(false);
+        }
+
+        @Override
+        public void onHeaders(Stream stream, HeadersFrame frame) {
+          if (frame.getMetaData() instanceof Response) {
+            responseCodeFuture.complete(((Response) frame.getMetaData()).getStatus());
+          }
+        }
+
+        @Override
+        public void onData(Stream stream, DataFrame frame, Callback callback) {
+          byte[] bytes = new byte[frame.getData().remaining()];
+          frame.getData().get(bytes);
+          resultBuilder.append(new String(bytes, StandardCharsets.UTF_8));
+
+          if (frame.isEndStream()) {
+            doneFuture.complete(true);
+          }
+
+          callback.succeeded();
+        }
+      };
+
+      FuturePromise<Stream> streamPromise = new FuturePromise<>();
+      session.newStream(headersFrame, streamPromise, responseListener);
+      Stream stream = streamPromise.get();
+
+      DataStream dataStream = new DataStream(stream, in);
+      try {
+        dataStream.send();
+      } catch (IOException e) {
+        if (!doneFuture.isDone()) {
+          throw e;
+        }
+      }
+
+      doneFuture.get();
+      client.stop();
+
+      int httpStatus = responseCodeFuture.get();
+      if (httpStatus != HttpStatus.OK_200) {
+        JSONObject responseJson = new JSONObject(resultBuilder.toString());
+        throw new DicomWebException("Http_" + responseCodeFuture.get()
+            + ", " + responseJson.getJSONObject("error").getString("status")
+            + ", " + responseJson.getJSONObject("error").getString("message"),
+            httpStatus, Status.ProcessingFailure);
+      }
+    } catch (Exception e) {
+      if (e instanceof DicomWebException) {
+        throw (DicomWebException) e;
+      }
+      throw new DicomWebException(e);
+    }
+  }
+
+  private static class DataStream {
```

Comment:
The Jetty implementation doesn't provide any utilities for streaming data? I would have thought there would be something we could use instead of having to write these lower level classes ourselves

Reply:
No, there doesn't seem to be a class for this.

There is whole high level Jetty client though (HttpClient class as opposed to low level Http2Client used by me). And it almost works, but ends up violating flow control during send more often than not (sending more data than allowed by WINDOW_UPDATE). I'm not sure how to fix this in high level client (If I did, I wouldn't need the low level implementation).

While current low level implementation just seems to work (I send next frame after callback succeeds, so no flow control errors there).

Reply:
Ok, can you please add a comment explaining why we need the low level method? (flow control error) Thanks